scripts/release/audit_rosdistro.py

```diff
@@ -0,0 +1,129 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import argparse
+import copy
+import pprint
+import sys
+
+from ros_buildfarm.argument import add_argument_arch
+from ros_buildfarm.argument import add_argument_build_name
+from ros_buildfarm.argument import add_argument_cache_dir
+from ros_buildfarm.argument import add_argument_config_url
+from ros_buildfarm.argument import add_argument_os_code_name
+from ros_buildfarm.argument import add_argument_os_name
+from ros_buildfarm.argument import add_argument_rosdistro_name
+from ros_buildfarm.common import filter_blocked_dependent_package_names
+from ros_buildfarm.common import get_cached_package_manifests
+from ros_buildfarm.common import get_os_package_name
+from ros_buildfarm.common import Target
+from ros_buildfarm.config import get_index as get_config_index
+from ros_buildfarm.config import get_release_build_files
+from ros_buildfarm.package_repo import get_package_repo_data
+
+from rosdistro import get_distribution_file
+from rosdistro import get_cached_distribution
+from rosdistro import get_index
+
+
+def main(argv=sys.argv[1:]):
+    parser = argparse.ArgumentParser(
+        description='Check if the sync criteria are matched to sync ' +
+                    'packages from the building to the testing repo')
+    add_argument_config_url(parser)
+    add_argument_rosdistro_name(parser)
+    add_argument_cache_dir(parser, '/tmp/package_repo_cache')
+    args = parser.parse_args(argv)
+
+    return run_audit(args.config_url, args.rosdistro_name, args.cache_dir)
+
+
+def run_audit(config_url, rosdistro_name, cache_dir):
+    config = get_config_index(config_url)
+    index = get_index(config.rosdistro_index_url)
+    dist_file = get_distribution_file(index, rosdistro_name)
+    build_files = get_release_build_files(config, rosdistro_name)
+    missing_packages = {}
+    for bf_name, bf_value in build_files.items():
+        missing_packages[bf_name] = copy.deepcopy(bf_value.targets)
+        for target in bf_value.get_targets_list():
+            all_pkgs, missing_pkgs = get_missing_packages(config_url, rosdistro_name, bf_name, target, cache_dir, deduplicate_dependencies=True)
+            missing_packages[bf_name][target] = missing_pkgs
+            if 'all' in missing_packages[bf_name]:
+                missing_packages[bf_name]['all'] &= missing_pkgs
+            else:
+                missing_packages[bf_name]['all'] = missing_pkgs
+
+            if 'all' in missing_packages:
+                missing_packages['all'] &= missing_pkgs
+            else:
+                missing_packages['all'] = missing_pkgs
+
+
+    recommended_actions = len(missing_packages['all'])
```

Comment:
A bit esoteric for a first review but if I don't mention it now I'll forget.
If there are exactly 256 missing packages (or 512, 768 or anything congruent to 0 modulo 255) this will have an exit code of 0 due to the overflow. What's the intent for using a semantically significant exit code?

Reply:
I changed it to return 1 if not zero recommendations.